scandium-core/src/main/java/org/eclipse/californium/scandium/CookieGenerator.java

```diff
@@ -41,55 +46,125 @@
+	/**
+	 * Mac initialized with {@link #lastSecretKey}, if clonable.
+	 */
 	private Mac hmac;
 
 	// attributes used for random byte generation
 	private final SecureRandom rng = new SecureRandom();
 	byte[] rd = new byte[32];
 
 	/**
+	 * Create HMAC.
+	 * 
+	 * Create HMAC cloning the master {@link #hmac}, if cloning is supported, or
+	 * create a new HAMC initialized with {@link #lastSecretKey}, if cloning is
+	 * not supported.
+	 * 
+	 * MUST be called in read or write scope of {@link #lock}!
+	 * 
+	 * @return new HMAC
+	 * @throws GeneralSecurityException if an security related exception occurs
+	 *             when creating the HMAC.
+	 */
+	private final Mac createHMAC() throws GeneralSecurityException {
+		if (hmac != null) {
+			// clone is supported
+			try {
+				return (Mac) hmac.clone();
+			} catch (CloneNotSupportedException e) {
+				throw new IllegalStateException("hmac doesn't support clone and MUST therefore be null!");
+			}
+		} else {
+			// clone is not supported
+			Mac hmac = Mac.getInstance("HmacSHA256");
+			hmac.init(lastSecretKey);
+			return hmac;
+		}
+	}
+
+	/**
 	 * Return a fresh HMAC algorithm instance.
 	 */
-	private Mac getHMAC() throws NoSuchAlgorithmException, CloneNotSupportedException, InvalidKeyException {
+	private Mac getHMAC() throws GeneralSecurityException {
+
 		lock.readLock().lock();
 		try {
-			if (hmac != null && !isKeyExpired()) {
-				return (Mac) hmac.clone();
+			// check, if a secret key is already created and not expired
+			if (lastSecretKey != null && !isKeyExpired()) {
+				return createHMAC();
 			}
 		} finally {
 			lock.readLock().unlock();
 		}
 
-		// if key expired or hmac not initialized;
+		// if key expired or secret key not initialized;
 		lock.writeLock().lock();
 		try {
 			// Recheck state because another thread might have acquired
 			// write lock and changed state before we did.
-			if (hmac == null) {
+			if (lastSecretKey == null) {
+				// initialize
+				generateSecretKey();
 				hmac = Mac.getInstance("HmacSHA256");
-				hmac.init(generateSecretKey());
+				hmac.init(lastSecretKey);
+				return (Mac) hmac.clone();
 			} else if (isKeyExpired()) {
-				hmac.init(generateSecretKey());
+				generateSecretKey();
 			}
-			return (Mac) hmac.clone();
+			return createHMAC();
+		} catch (CloneNotSupportedException ex) {
+			// catch, if MacSpi is not cloneable (android)!
+			Mac hmac = this.hmac;
+			this.hmac = null;
+			return hmac;
 		} finally {
 			lock.writeLock().unlock();
 		}
 	}
 
+	/**
+	 * check, if {@link #lastSecretKey} has expired.
+	 * 
+	 * MUST be called in read or write scope of {@link #lock}!
+	 * 
+	 * @return {@code true}, if key is expired, {@code false} otherwise.
+	 */
 	private boolean isKeyExpired() {
-		return System.currentTimeMillis() - lastGenerationDate > keyLifetime;
+		// consider sign wrap in longs (very optimistic about the uptime :-) )
+		return (System.nanoTime() - nextKeyGenerationNanos) >= 0;
```

Comment:
I have no clue if the JVM actually does what you have written explicitly when you want to check
`System.nanoTime() >= nextKeyGenerationNanos`
or if the comparison can be done more efficiently at the bit level. But this seems to be an academic discussion ;-)

Reply:
Javadoc of nanoTime recommends this :-).

https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#nanoTime()

> To compare two nanoTime values
>
> long t0 = System.nanoTime();
> ...
> long t1 = System.nanoTime();
>
> one should use t1 - t0 < 0, not t1 < t0, because of the possibility of numerical overflow.